Firestore/core/src/util/thread_safe_memoizer.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_UTIL_THREAD_SAFE_MEMOIZER_H_
+#define FIRESTORE_CORE_SRC_UTIL_THREAD_SAFE_MEMOIZER_H_
+
+#include <mutex>  // NOLINT(build/c++11)
+#include <vector>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * Stores a memoized value in a manner that is safe to be shared between
+ * multiple threads.
+ */
+template <typename T>
+class ThreadSafeMemoizer {
```

Comment:
My last thought is that we should make `ThreadSafeMemoizer` copyable, so that the usages don't have to do that ugly hack of wrapping it in a std::shared_ptr. But let's do that in a follow-up PR.

Reply:
After thinking about it more, I think the work of making `ThreadSafeMemoizer` movable and copyable should just be put in the backlog. Open a buganizer ticket and put a "TODO(b/NNNNNN) Make copyable and moveable" comment above the deleted copy/move method.